File: python/tests/test_writer.py
# This file is part of an MIT-licensed project: see LICENSE file or README.md for details.
# Copyright (c) 2025 Ian Thomas

import unittest
import sys
import os

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../src")))

from expression_parser.parser import Parser
from expression_parser.writer import Writer, STRING_FORMAT_SINGLEQUOTE, STRING_FORMAT_ESCAPED_SINGLEQUOTE, STRING_FORMAT_DOUBLEQUOTE, STRING_FORMAT_ESCAPED_DOUBLEQUOTE

class TestWriter(unittest.TestCase):

    def setUp(self):
        self.maxDiff = None  # Allow full diff output for every test case

    def _load_file(self, file_name):
        try:
            with open(f"../tests/{file_name}", "r", encoding="utf-8") as file:
                return file.read()
        except Exception as e:
            self.fail(f"Error loading {file_name}: {e}")

    def test_simple(self):
        
        parser = Parser()
        expression = parser.parse("get_name()=='fred' and counter>0 and 5/5.0!=0")

        result = expression.write()
        self.assertEqual(result, "get_name() == 'fred' and counter > 0 and 5 / 5 != 0", "Expression doesn't match.")
        Writer.string_format = STRING_FORMAT_DOUBLEQUOTE
        result = expression.write()
        self.assertEqual(result, "get_name() == \"fred\" and counter > 0 and 5 / 5 != 0", "Expression doesn't match.")
        Writer.string_format = STRING_FORMAT_ESCAPED_DOUBLEQUOTE
        result = expression.write()
        self.assertEqual(result, "get_name() == \\\"fred\\\" and counter > 0 and 5 / 5 != 0", "Expression doesn't match.")
        Writer.string_format = STRING_FORMAT_ESCAPED_SINGLEQUOTE
        result = expression.write()
        self.assertEqual(result, "get_name() == \\'fred\\' and counter > 0 and 5 / 5 != 0", "Expression doesn't match.")
        Writer.string_format = STRING_FORMAT_SINGLEQUOTE

    def test_writer(self):

        source = self._load_file("Writer.txt")

        # Split into lines
        lines = source.splitlines()

        parser = Parser()

        processed_lines = []
        for line in lines:
            if (line.startswith("//")):
                processed_lines.append(line)
                continue

            processed_lines.append(f'"{line}"')

            node = parser.parse(line)
            
            processed_lines.append(node.write())

            processed_lines.append("")

        output = "\n".join(processed_lines)

        #print(output)

        match = self._load_file("Writer-Output.txt")
        self.assertMultiLineEqual(match, output)

File: cpp/test/test_writer.cpp
// This file is part of an MIT-licensed project: see LICENSE file or README.md for details.
// Copyright (c) 2025 Ian Thomas

#include "expression_parser/parser.h"
#include "expression_parser/writer.h"
#include "catch_amalgamated.hpp"
#include "test_utils.h"
#include <fstream>

using namespace ExpressionParser;

TEST_CASE( "SimpleWriter") {
    Parser parser;
    auto expression = parser.Parse("get_name()=='fred' and counter>0 and 5/5.0!=0");

    std::string result = expression->Write();
    REQUIRE(result == "get_name() == 'fred' and counter > 0 and 5 / 5 != 0");

    Writer::setStringFormat(STRING_FORMAT::DOUBLEQUOTE);
    result = expression->Write();
    REQUIRE(result == "get_name() == \"fred\" and counter > 0 and 5 / 5 != 0");

    Writer::setStringFormat(STRING_FORMAT::ESCAPED_DOUBLEQUOTE);
    result = expression->Write();
    REQUIRE(result == "get_name() == \\\"fred\\\" and counter > 0 and 5 / 5 != 0");

    Writer::setStringFormat(STRING_FORMAT::ESCAPED_SINGLEQUOTE);
    result = expression->Write();
    REQUIRE(result == "get_name() == \\'fred\\' and counter > 0 and 5 / 5 != 0");

    Writer::setStringFormat(STRING_FORMAT::SINGLEQUOTE);
}

TEST_CASE( "MatchOutputWriter") {
    std::string source = loadTestFile("Writer.txt");

    std::vector<std::string> lines;
    std::istringstream iss(source);
    std::string line;
    while (std::getline(iss, line)) {
        lines.push_back(line);
    }

    Parser parser;
    std::vector<std::string> processedLines;
    for (const auto &l : lines) {
        if (l.rfind("//", 0) == 0) {
            processedLines.push_back(l);
            continue;
        }

        // Append the original line wrapped in quotes.
        processedLines.push_back("\"" + l + "\"");
        try {
            auto node = parser.Parse(l);
            processedLines.push_back(node->Write());
        }
        catch (const std::exception &ex) {
            processedLines.push_back(ex.what());
        }
        processedLines.push_back("");
    }

    std::string output = joinStrings(processedLines, "\n");

    /*std::ofstream outFile("output.txt");
    if (!outFile) {
        throw std::runtime_error("Failed to open output.txt for writing.");
    }
    outFile << output;
    outFile.close();*/

    std::string match = loadTestFile("Writer-Output.txt");
    REQUIRE(match == output);
}